### PR TITLE
Modify slackrtm so that it forwards messages posted by the bot

### PR DESCRIPTION
**Pull request by neon-ninja**
Also execute bot commands triggered from slack

### CODE REVIEW

**Comment by thannemann** [comment]
This causes messages from slack to echo back into the slack chat when it posts in Hangouts

**Comment by neon-ninja** [comment]
@thannemann ^ patched

**Comment by thannemann** [comment]
I am still seeing the echo back is anyone else?

**Comment by phantomdarkness** [comment]
No, working for me with no echo so far.

**Comment by phantomdarkness** [comment]
@neon-ninja bot is echoing images sent from slack.

**Comment by neon-ninja** [comment]
@thannemann @phantomdarkness ^ try that, think I've fixed both issues

**Comment by phantomdarkness** [comment]
2 quick tests, seems to be working with images.

**Comment by thannemann** [comment]
I am still seeing random echos, but this time I happened to notice the following error just before it happened.

20:44:27 ERROR plugins.slackrtm: _handle_slackout threw: list index out of range
Traceback (most recent call last):
  File "/home/XXXXXX/hangoutsbot/hangupsbot/plugins/slackrtm.py", line 1365, in _handle_slackout
    yield from slackrtm.handle_ho_message(event)
  File "/home/XXXXXX/hangoutsbot/hangupsbot/plugins/slackrtm.py", line 1175, in handle_ho_message
    command = event.text.split(': ')[1]
IndexError: list index out of range

**Comment by neon-ninja** [comment]
@thannemann echos seemed to be caused when a slack channel was being spammed. I think it's cause there may have been network lag, causing hangouts messages that were sent in quick succession to appear after following messages had been sent. try the above fix and see if that works for you. thanks for helping test this PR :)

**Comment by thannemann** [comment]
not sure if it's related, but I am getting the following errors, I am going to roll back and test a unmodified version

  File "/home/XXXXXX/hangoutsbot/hangupsbot/plugins/slackrtm.py", line 602, in handleCommands
    userID = self.get_slackDM(msg.user)
  File "/home/XXXXX/hangoutsbot/hangupsbot/plugins/slackrtm.py", line 389, in get_slackDM
    self.dminfos[userid] = self.api_call('im.open', user = userid)['channel']
TypeError: byte indices must be integers, not str

**Comment by thannemann** [comment]
no not related.

**Comment by pleasantone** [comment]
Testers, what is the state of this fix?  Is it working perfectly or still buggy?

File: hangupsbot/plugins/slackrtm.py [hunk 1]
```diff
@@ -1157,13 +1159,28 @@ def handle_reply(self, reply):
...
                     else:
                         # we should not upload the images, so we have to send the url instead
                         response += msg.file_attachment
+                self.sending += 1
                 self.bot.send_html_to_conversation(sync.hangoutid, response)
 
+    @asyncio.coroutine
     def handle_ho_message(self, event):
         for sync in self.get_syncs(hangoutid=event.conv_id):
+            if self.sending and ': ' in event.text:
```
Comment (pleasantone) [inline hangupsbot/plugins/slackrtm.py]: This is a pretty horrible way to test for that.  Can we use the message context instead?

**Comment by phantomdarkness** [comment]
I've seen a few random dupe messages but they are few and far between. Overall it ls been working just fine

**Comment by pleasantone** [comment]
OK, I'm happy to merge it once it's working properly, but "a few dupe message" sounds like it's not yet ready for staging, which bums me out, because I'd love to see this functionality in the bot.

**Comment by phantomdarkness** [comment]
I'm not sure if the dupes are related to this pull or just slack in general. I think it's good to merge

**Comment by pleasantone** [comment]
Hrm... I would really hate to start a meltdown somewhere... or we could just say "it's a good day to die, ship it!"

With Arthur, Nihal, and I all focusing on other things right now, I'm inclined to be conservative until more people chime in and say they've been testing it and have had no issues.  There should never be a duplicate message, the only time I've ever seen one in the mainline code was during a bot restart in the middle of syncing.

**Comment by phantomdarkness** [comment]
I'm thinking that might be the cause of the dupe. Others can chime in with their experience. I've loved it so far and run it in production

File: hangupsbot/plugins/slackrtm.py (hunk 1 above)
Comment (neon-ninja) [inline hangupsbot/plugins/slackrtm.py]: Easier said than done it seems.. that'd require refactoring self.bot.send_html_to_conversation to use coro_send_message, something I can't seem to get to work

Comment (neon-ninja) [inline hangupsbot/plugins/slackrtm.py]: Ok I managed to add a key to the context on send with `self.bot.coro_send_message(sync.hangoutid, response, context={'slack': True}))` - but how do I get that out of the event?

**Comment by thiasB** [comment]
I noticed one dupe today

**Comment by pleasantone** [comment]
@neon-ninja Nick, I'm prepared to merge it, unfortunately, you based the pull request off of master instead of staging. Can you please rebase it?

**Comment by azack** [comment]
I'm not sure what about my environment is different, but with this configured with slackrtm syncing two hangouts, images posted to one hangout will cause an infinite loop of images between the two hangouts. It appears as if it doesn't recognize that the image was posted by a sync'd message and repeats it again.

**Comment by neon-ninja** [comment]
@azack try comment out https://github.com/neon-ninja/hangupsbot/blob/1090e29d059247f83de3b15ab3631ab55faea9fc/hangupsbot/plugins/slackrtm.py#L1184 and see if that fixes your specific issue. Although in theory, this could cause a bug where if a slack user posts the same image twice, without a different image in between, the second image might not be posted into hangouts.

**Comment by azack** [comment]
@neon-ninja There is no loop with desktop images, but it does appear with android-posted images, even with that line commented out.  Reported as #613

**Pull request merged**